File: 2018/day4.nim
from algorithm import sort
from re import nil
from strUtils import endsWith, parseInt, split
import tables
from times import nil

proc step1() =
  var rows: seq[string]
  for row in lines("day4input.txt"):
    rows.add(row)
  sort(rows, cmp)

  var minutes_asleep_per_guard = newCountTable[int]()
  var sleepiest_mins = initTable[int, array[60, int]]()
  var sleeping_since = times.now()
  var current_guard: int

  for row in rows:
    var matches: array[2, string]
    discard re.match(row, re.re"\[([\d\s:-]+)\] (.+)$", matches)
    let date = times.parse(matches[0], "yyyy-MM-dd HH:mm")

    if matches[1].endsWith("begins shift"):
      current_guard = parseInt(split(matches[1])[1][1..^1])
    elif matches[1].endsWith("falls asleep"):
      sleeping_since = date
    else: # Woke up
      let slept_minutes = times.between(sleeping_since, date).minutes

      if not minutes_asleep_per_guard.hasKey(current_guard):
        minutes_asleep_per_guard[current_guard] = slept_minutes
      else:
        inc(minutes_asleep_per_guard[current_guard], slept_minutes)

      var default_sleepiest_mins: array[60, int]
      if not sleepiest_mins.hasKey(current_guard):
        sleepiest_mins[current_guard] = default_sleepiest_mins

      if sleeping_since.minute > date.minute:
        for m in sleeping_since.minute..59:
          inc(sleepiest_mins[current_guard][m])
        for m in 0..<date.minute:
          inc(sleepiest_mins[current_guard][m])
      else:
        for m in sleeping_since.minute..<date.minute:
          inc(sleepiest_mins[current_guard][m])

  let most_asleep = minutes_asleep_per_guard.largest
  echo "The sleepiest guard ", most_asleep[0], " slept in total for ", most_asleep[1], " mins"

  # Step 2:
  var
    max_frequent_guard_id = most_asleep[0]
    max_frequent_minute = 0
  for guard_id, sleep_minutes in sleepiest_mins:
    for i, m in sleep_minutes:
      if m > sleepiest_mins[max_frequent_guard_id][max_frequent_minute]:
        max_frequent_guard_id = guard_id
        max_frequent_minute = i

  echo "Most frequent minute was for guard ", max_frequent_guard_id, ", minute ", max_frequent_minute


when isMainModule:
  var time = times.cpuTime()
  step1()
  echo "Steps 1 and 2 took: ", (times.cpuTime() - time) * 1000, " ms"

File: 2018/day4.py
from collections import defaultdict
import datetime
import time

def steps_1_and_2_b():
    rows = []
    with open("day4input.txt", "r") as input_f:
        rows = input_f.readlines()
    rows.sort()

    guards = defaultdict(lambda: 60 * [0]) # TODO try with dict instead of list
    current_guard = None
    sleeping_since = None

    for row in rows:
        parts = row.split()
        date_str = (parts[0] + parts[1])[1:-1]
        date = datetime.datetime.strptime(date_str, '%Y-%m-%d%H:%M')

        if parts[-1] == 'shift':
            # New guard started a shift:
            current_guard = int(parts[3][1:])
        elif parts[-1] == "asleep":
            # Guard fell asleep:
            sleeping_since = date
        else:
            # Guard woke up:
            if sleeping_since.minute > date.minute:
                for m in range(sleeping_since.minute, 60):
                    guards[current_guard][m] += 1
                for m in range(date.minute):
                    guards[current_guard][m] += 1
            else:
                for m in range(sleeping_since.minute, date.minute):
                    guards[current_guard][m] += 1

    top_sleeper = None
    max_sleeper_mins = 0
    top_frequency_sleeper = None
    max_freq_min = 0
    max_freq_mins_slept = 0

    for guard, minutes in guards.items():
        sleeper_mins = sum(minutes)
        if sleeper_mins > max_sleeper_mins:
            top_sleeper = guard
            max_sleeper_mins = sleeper_mins
        for i, m in enumerate(minutes):
            if m > max_freq_mins_slept:
                max_freq_mins_slept = m
                max_freq_min = i
                top_frequency_sleeper = guard

    print("The sleepiest guard", top_sleeper, "slept in total for", max_sleeper_mins, "mins")
    print("Most frequent minute was for guard", top_frequency_sleeper, ", minute", max_freq_min)


if __name__ == "__main__":
    t = time.time()
    steps_1_and_2_b()
    print("Steps 1 and 2 took:", round((time.time() - t) * 1000, 2), "ms")
